File: civitasJAVA/src/civitas/SorpresaSalirCarcel.java
package civitas;

import java.util.ArrayList;

/**
 * @author Jesús Navarro Merino
 * @author José Alberto Gómez García
 */

public class SorpresaSalirCarcel extends Sorpresa {

    private MazoSorpresas mazo;

    public SorpresaSalirCarcel(String texto, MazoSorpresas _mazo) {
        super(texto);
        mazo=_mazo;
    }

    @Override
    void aplicarAJugador(int actual, ArrayList<Jugador> todos) {
        boolean hay_s = false;
        
        if(jugadorCorrecto(actual,todos)){
            informe(actual,todos);
            //preguntar si alguien tiene salvonducto
            for(int i=0; i<todos.size() && !hay_s; i++){
                if(todos.get(i).tieneSalvoconducto()) {
                    hay_s = true;  
                }
            }
           if(!hay_s){
               (todos.get(actual)).obtenerSalvoconducto(this);
               this.salirDelMazo(); 
           }
        }
    }
    
    @Override
    protected void informe (int actual, ArrayList<Jugador> todos){
        super.informe(actual, todos);
    }

    void salirDelMazo(){
        mazo.inhabilitarCartaEspecial(this);
    }

    void usada (){
        mazo.habilitarCartaEspecial(this);
    }
    
}

File: civitasRUBY/lib/sorpresa_salir_carcel.rb
# encoding:utf-8

module Civitas
  class SorpresaSalirCarcel < Sorpresa
    
    attr_reader :texto, :carcel
    
    def initialize(texto,_mazo)
      super(texto)
      @mazo=_mazo      
    end
    
    def self.constructorSorpresaSalirCarcel(texto,mazo)
      new(texto,mazo)
    end
    
    def aplicarAJugador(actual,todos)
      hay_s = false
        
        if(jugadorCorrecto(actual, todos))
          informe(actual, todos)
          i = 0
          while (i < todos.size and !hay_s)
            if (todos[i].tieneSalvoconducto)
              hay_s = true
            end
            i = i+1
          end
          
          if(!hay_s)
            todos[actual].obtenerSalvoconducto(self)
            salirDelMazo
          end
          
        end
      
    end
    
    def informe(actual,todos)
      super(actual,todos)
    end
    
    def salirDelMazo()
      @mazo.inhabilitarCartaEspecial(self)
    end
    
    def usada()
      @mazo.habilitarCartaEspecial(self)
    end
    
    public_class_method :new
    
  end
end
